File: 03852. Smallest Pair With Different Frequencies/solution.ts
/**
 * Problem: 3852. Smallest Pair With Different Frequencies
 *
 * Difficulty: Easy
 *
 * Language: TypeScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Finds pair of numbers with different frequencies
 *
 * @param nums - Input array of integers
 *
 * @returns Two numbers with different frequencies or [-1, -1]
 */
const minDistinctFreqPair = (nums: number[]): number[] => {
  // Frequency array for numbers 1-100 (index 0 unused)
  const frequency: number[] = new Array(101).fill(0)

  // Count occurrences of each number
  for (const num of nums) frequency[num]++

  // Track first number found with frequency
  let firstNumber: number = -1

  // Iterate through all possible numbers
  for (let i: number = 0; i < frequency.length; i++) {
    // Check if number exists in array
    if (frequency[i] > 0) {
      // If we already found a first number, compare frequencies
      if (firstNumber !== -1) {
        // If frequencies differ, return the pair
        if (frequency[firstNumber] !== frequency[i]) return [firstNumber, i]
      }
      // Otherwise, set as first number
      else firstNumber = i
    }
  }

  // No valid pair found
  return [-1, -1]
}

File: 03852. Smallest Pair With Different Frequencies/solution.cs
/**
 * Problem: 3852. Smallest Pair With Different Frequencies
 *
 * Difficulty: Easy
 *
 * Language: C#
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

public class Solution
{
  public int[] MinDistinctFreqPair(int[] nums)
  {
    // Allocate frequency array on stack for performance (numbers 1-100)
    Span<int> frequencies = stackalloc int[101];

    // Count occurrences of each number
    foreach (var x in nums)
      frequencies[x]++;

    // Get array length
    var n = nums.Length;

    // Iterate through each number
    for (var x = 1; x < 101; x++)
    {
      // Get frequency of current number
      var freqX = frequencies[x];

      // Skip if number doesn't exist in array
      if (0 == freqX)
        continue;

      // Compare with all subsequent numbers
      for (var y = x + 1; y < 101; y++)
      {
        // Get frequency of second number
        var freqY = frequencies[y];

        // Skip if number doesn't exist or frequencies are equal
        if (0 == freqY || freqX == freqY)
          continue;

        // Found pair with different frequencies
        return [x, y];
      }
    }

    // No valid pair found
    return [-1, -1];
  }
}
